pkgs/by-name/vc/vcpkg/package.nix

```diff
@@ -41,6 +41,7 @@ stdenvNoCC.mkDerivation (finalAttrs: {
       ''}
 
       ln -s "$out/bin/vcpkg" "$out/share/vcpkg/vcpkg"
+      echo '{ "readonly": true, "usegitregistry": true, "deployment": "Git", "embeddedsha": "f7423ee180c4b7f40d43402c2feb3859161ef625" }' > "$out/share/vcpkg/vcpkg-bundle.json"
```

Comment:
Could you add an explanation or link to documentation of vcpkg-bundle.json?

Reply:
When asking Billy O'Neal about documentation for that file, he answered me that there's no documentation for it. Given that this file is intended to be written by those who ship vcpkg, and that the only official distributions of vcpkg right now is by microsoff, I could see why they haven't taken the time to document it.

He gave me a link of those two source file as help for implementing this PR though:

 - https://github.com/microsoft/vcpkg-tool/blob/f098d3e0aaa7e46ea84a1f7079586e1ec5af8ab5/vcpkg-init/mint-standalone-bundle.ps1#L21
 - https://github.com/microsoft/vcpkg-tool/blob/f098d3e0aaa7e46ea84a1f7079586e1ec5af8ab5/src/vcpkg/bundlesettings.cpp#L87

I will add links to the relevant place in the source code right above the echo

Reply:
I'm not sure I understand what the `sha` represents. Is it the `sha` of vcpkg.exe?

Reply:
I think I found it https://github.com/microsoft/vcpkg-tool/blob/d272c0d4f5175b26bd56c6109d4c4935b791a157/src/vcpkg/commands.new.cpp#L190

So it's the sha of the current vcpkg library, if I understand correctly

Reply:
The sha of the current registry, which is simply the current git hash of the vcpkg repo.

Reply:
Nix can't do `src.currentGitHash` or something like that, right?

Reply:
> Nix can't do `src.currentGitHash` or something like that, right?

Well, the currentGitHash is simply the revision, isn't it?

Reply:
I think we could use replace the tag with the commit hash in `src.rev`, and create an update script that makes sure to resolve the next stable tag's commit hash and set that instead of the tag. Normally it wouldn't make sense since the tag is a better revision to target, but here we have an actual use case for it.

> When asking Billy O'Neal about documentation for that file, he answered me that there's no documentation for it. Given that this file is intended to be written by those who ship vcpkg, and that the only official distributions of vcpkg right now is by microsoff, I could see why they haven't taken the time to document it.

Do you know whether we have any guarantees (or at least hopes) of stability here? Or could this interface change at any time? It would be nice not to have to rethink this solution in the near future, in case the inner workings are changed up.

Reply:
@gracicot

> Nix can't do `src.currentGitHash` or something like that, right?

It looks like it can:

```diff
diff --git a/pkgs/by-name/vc/vcpkg/package.nix b/pkgs/by-name/vc/vcpkg/package.nix
index b9e51af8bc8b..8e1a668139f1 100644
--- a/pkgs/by-name/vc/vcpkg/package.nix
+++ b/pkgs/by-name/vc/vcpkg/package.nix
@@ -1,5 +1,4 @@
-{ fetchFromGitHub
-, stdenvNoCC
+{ stdenvNoCC
 , lib
 , vcpkg-tool
 , makeWrapper
@@ -10,11 +9,9 @@ stdenvNoCC.mkDerivation (finalAttrs: {
   pname = "vcpkg";
   version = "2024.06.15";
 
-  src = fetchFromGitHub {
-    owner = "microsoft";
-    repo = "vcpkg";
-    rev = finalAttrs.version;
-    hash = "sha256-eDpMGDtC44eh0elLWV0r1H/WbpVdZ5qMedKh7Ct50Cs=";
+  src = builtins.fetchGit {
+    url = "https://github.com/microsoft/vcpkg";
+    ref = "refs/tags/${finalAttrs.version}";
   };
 
   nativeBuildInputs = [ makeWrapper ];
@@ -50,7 +47,7 @@ stdenvNoCC.mkDerivation (finalAttrs: {
       # will checkout a remote instead of trying to do git operation in the vcpkg root
       #   https://github.com/microsoft/vcpkg-tool/blob/d272c0d4f5175b26bd56c6109d4c4935b791a157/src/vcpkg/vcpkgpaths.cpp#L920
       #   https://github.com/microsoft/vcpkg-tool/blob/d272c0d4f5175b26bd56c6109d4c4935b791a157/src/vcpkg/configuration.cpp#L718
-      echo '{ "readonly": true, "usegitregistry": true, "deployment": "Git", "embeddedsha": "f7423ee180c4b7f40d43402c2feb3859161ef625" }' > "$out/share/vcpkg/vcpkg-bundle.json"
+      echo '{ "readonly": true, "usegitregistry": true, "deployment": "Git", "embeddedsha": "${finalAttrs.src.rev}" }' > "$out/share/vcpkg/vcpkg-bundle.json"
       touch "$out/share/vcpkg/vcpkg.disable-metrics"
 
       runHook postInstall
```

Reply:
I saw that `leaveDotGit` uses `builtin.fetchGit` internally. I changed it to use git to get the value of embeddedsha

Reply:
@Jayman2000 Let me know if you prefer using fetchGit directly instead

Reply:
> @Jayman2000 Will the script that generate the automatic update still work if we switch to `builtin.fetchGit`? Such as these: #327790

I don’t really know. Pull requests like that one are generated by [`nixpkgs-update`](https://github.com/nix-community/nixpkgs-update). Normally, I would just run `nixpkgs-update` on my local system in order to see if it works, but I can’t because of nix-community/nixpkgs-update#335. That being said, there is another tool with a confusingly similar name: [`nix-update`](https://github.com/Mic92/nix-update). As you saw in NixOS/nix#11222, `nix-update` doesn’t work if you use `builtins.fetchGit`. If `nix-update` doesn’t work, then I doubt that `nixpkgs-update` will work.

> I saw that `leaveDotGit` uses `builtin.fetchGit` internally. I changed it to use git to get the value of embeddedsha
>
> @Jayman2000 Let me know if you prefer using fetchGit directly instead

Here’s my concern: if we use `leaveDotGit`, then won’t we run into #8567 again?

[Here’s a potential fixup for this PR.](https://github.com/Jayman2000/nixpkgs-pr/commit/9a11bce3eb0ddb7196cc24ca357091d20d88fe36) I kept `leaveDotGit = true`, but I delete the `.git` folder before the fetcher calculates the hash. Hopefully, that will prevent the hash from changing under our feet without breaking `nixpkgs-update`.

Reply:
@Jayman2000 I applied the fixup, seems to work great